# Contoh Aplikasi Aljabar Linier

## Algoritma PageRank
Algoritma PageRank adalah sebuah algoritma yang digunakan oleh mesin pencari Google untuk menentukan peringkat halaman web dalam hasil pencarian mereka. Algoritma ini dikembangkan oleh Larry Page dan Sergey Brin pada tahun 1996 saat mereka masih mahasiswa di Universitas Stanford. Ini pertama kali digunakan untuk menentukan peringkat halaman web di mesin pencari Google. Saat ini, semakin banyak digunakan di berbagai bidang, misalnya dalam menentukan peringkat pengguna di media sosial dll

## Penerapan PageRank Menggunakan Python

In [1]:
import numpy as np

def perkalianMatrix(A, V, conv=0.01, itr=0):
    # Mengalikan matriks A dengan vektor V
    A_ = A.dot(V)
    
    # Memeriksa apakah konvergen (perubahan kurang dari ambang batas)
    if np.linalg.norm(A_ - V) < conv:
        return A_, itr
    else:
        # Jika belum konvergen, lanjutkan iterasi dengan matriks baru
        return perkalianMatrix(A, A_, conv, itr+1)

# Matriks A (matriks tautan)
A = np.array([[0, 0, 0, 0, 1/3],
              [0, 0, 1/2, 1, 1/3],
              [1/3, 0, 0, 0, 1/3],
              [1/3, 0, 1/2, 0, 0],
              [1/3, 1, 0, 0, 0]])

# Vektor PageRank awal (semua halaman memiliki nilai awal 0.2)
V = np.array([0.2, 0.2, 0.2, 0.2, 0.2])

# Hitung PageRank
hasil, iterasi = perkalianMatrix(A, V)

# Tampilkan hasil
print("Hasil konvergen:")
print(hasil)
print()
print("Jumlah iterasi:", iterasi)
print()

# Urutkan nilai PageRank
hasil_terurut = sorted(hasil, reverse=True)
print("Nilai PageRank yang diurutkan:")
for i, pr in enumerate(hasil_terurut):
    print(f"Halaman {i}: {pr:}")

Hasil konvergen:
[0.11150739 0.293549   0.14974089 0.11342783 0.33177488]

Jumlah iterasi: 7

Nilai PageRank yang diurutkan:
Halaman 0: 0.3317748818777625
Halaman 1: 0.29354900167657366
Halaman 2: 0.149740893156531
Halaman 3: 0.11342783112330437
Halaman 4: 0.11150739216582836


## Penjelasan

1. **Hasil Konvergen**:

   Hasil konvergen adalah:

   \[ $\text{PageRank} = \begin{bmatrix}
   0.331 \\
   0.293 \\
   0.149 \\
   0.113 \\
   0.331 \\
   \end{bmatrix}$ \]

2. **Nilai PageRank yang Diurutkan**:

   Nilai PageRank yang diurutkan dari yang tertinggi ke terendah:

   - Halaman 1: 0.331
   - Halaman 2: 0.293
   - Halaman 3: 0.149
   - Halaman 4: 0.113
   - Halaman 5: 0.111